Move THCTensor_(uniform) to ATen

syed-ahmed opened the pull request:
# Summary
As a first step for this plan: https://github.com/pytorch/pytorch/issues/19508#issuecomment-485178192, this PR moves `THCTensor_(uniform)` to ATen. Major changes are:
- `uniform_` cuda kernel now utilizes a philox generator.
- the kernel also utilizes TensorIterator
- the kernel uses a grid-stride loop to achieve peak effective bandwidth

# BC breaking change
- Since the engine has changed from `curandStateMTGP32` to `curandStatePhilox4_32_10`, the randoms generated now will be different.
- Here is the diff showing codegen changes: https://gist.github.com/syed-ahmed/4af9ae0d42b6c7dbaa13b9dd0d1dd1e8 (BC breaking change if any)

# Testing
- Philox4_32_10 is known to pass the standard TestU01 Big Crush test (https://www.thesalmons.org/john/random123/papers/random123sc11.pdf) and hence the quality of random numbers generated isn't an issue when compared to the previously used `curandStateMTGP32`.
- I have added a test case in `aten/src/ATen/test/cuda_distributions_test.cu` which verifies that philox offset is incremented properly

# Benchmark
The benchmark was done on a DGX station with 4 V100s.
## Runtime
I modified the script from @jcjohnson 's [multinomial benchmark](https://github.com/jcjohnson/pytorch-multinomial-benchmark) to produce this notebook which shows that there is a general speedup with this PR and a regression hasn't been introduced: https://gist.github.com/syed-ahmed/9d26d4e96308aed274d0f2c7be5218ef

To reproduce the notebook:
- Run https://gist.github.com/syed-ahmed/4208c22c541f1d30ad6a9b1efc1d728f in a container with the current pytorch top of tree with the command: `python uniform_benchmark.py --stats_json before.json`
- Apply this diff to the current pytorch top of tree and run the same script in a container with the command: `python uniform_benchmark.py --stats_json after.json`
- Run the notebook attached above with the `after.json` and `before.json` in the same directory

## Effective Bandwidth
The effected bandwidth was calculated using the script (thanks to @ngimel ): https://gist.github.com/syed-ahmed/f8b7384d642f4bce484228b508b4bc68
Following are the numbers before and after.
#### Before
```
uniform, size, elements 65536 forward 5.168914794921875e-06 bandwidth (GB/s) 50.71548098597786
uniform, size, elements 131072 forward 5.056858062744141e-06 bandwidth (GB/s) 103.67860705101367
uniform, size, elements 262144 forward 7.164478302001953e-06 bandwidth (GB/s) 146.357621001797
uniform, size, elements 524288 forward 1.1217594146728515e-05 bandwidth (GB/s) 186.9520302275877
uniform, size, elements 1048576 forward 1.923084259033203e-05 bandwidth (GB/s) 218.10297600317384
uniform, size, elements 2097152 forward 3.640890121459961e-05 bandwidth (GB/s) 230.39992200138826
uniform, size, elements 4194304 forward 6.778717041015625e-05 bandwidth (GB/s) 247.49839679819922
uniform, size, elements 8388608 forward 0.00012810707092285157 bandwidth (GB/s) 261.92490202361347
uniform, size, elements 16777216 forward 0.00025241613388061524 bandwidth (GB/s) 265.86598474620627
uniform, size, elements 33554432 forward 0.000497891902923584 bandwidth (GB/s) 269.5720239913193
```
#### After
```
uniform, size, elements 65536 forward 5.550384521484375e-06 bandwidth (GB/s) 47.22988091821306
uniform, size, elements 131072 forward 5.581378936767578e-06 bandwidth (GB/s) 93.93520954942333
uniform, size, elements 262144 forward 6.165504455566406e-06 bandwidth (GB/s) 170.071404141686
uniform, size, elements 524288 forward 6.3276290893554685e-06 bandwidth (GB/s) 331.4277702414469
uniform, size, elements 1048576 forward 8.509159088134765e-06 bandwidth (GB/s) 492.91639239047356
uniform, size, elements 2097152 forward 1.2989044189453124e-05 bandwidth (GB/s) 645.8218077979443
uniform, size, elements 4194304 forward 2.347707748413086e-05 bandwidth (GB/s) 714.6211452997259
uniform, size, elements 8388608 forward 4.4286251068115234e-05 bandwidth (GB/s) 757.6715389250498
uniform, size, elements 16777216 forward 8.672237396240235e-05 bandwidth (GB/s) 773.8356427961071
uniform, size, elements 33554432 forward 0.00016920566558837892 bandwidth (GB/s) 793.2224227438523
```

syed-ahmed committed:
initial commit

syed-ahmed committed:
Change philox offset calc

syed-ahmed committed:
Add static_assert

syed-ahmed commented:
@gchanan I have used the guts of your `retainTensorImpl` function to be able to call at::native here: https://github.com/pytorch/pytorch/pull/20292/files#diff-33ca813db2486ebd0c41ed5d96658bfdR179.

@cpuhrsch the codegen changes are attached in the PR description.

ezyang commented on aten/src/THC/generic/THCTensorRandom.cu:
```diff
@@ -380,8 +367,12 @@ void THCTensor_(multinomialAliasDraw)(THCState *state, THCudaLongTensor *self, T
+  c10::raw::intrusive_ptr::incref(uniform);
+  c10::raw::intrusive_ptr::incref(bernoulli);
+  auto out_uniform = at::Tensor(c10::intrusive_ptr<at::TensorImpl, at::UndefinedTensorImpl>::reclaim(uniform));
+  auto out_bernoulli = at::Tensor(c10::intrusive_ptr<at::TensorImpl, at::UndefinedTensorImpl>::reclaim(bernoulli));
```
You should just use `THTensor_wrap` here

ezyang commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -486,6 +486,13 @@ std::unique_ptr<TensorIterator> TensorIterator::unary_op(Tensor& out, const Tens
   return builder.build();
 }
 
+std::unique_ptr<TensorIterator> TensorIterator::nullary_op(Tensor& out) {
```
Umm... are you sure? Isn't this just an in-place unary op?

cc @colesbury

colesbury replied:
It takes no Tensor arguments so it's nullary. `f() -> [random]`. As opposed to something like `x.pow_(2)` which is the inplace version of a unary function: `f(x) -> x^2`. (The naming is unfortunate and I'm probably playing fast and loose with the terminology, but... oh well)

Since it doesn't read `out` it shouldn't be added as an input below.

syed-ahmed replied:
I did try not adding `out` as an input, but got the weird behavior below where I keep getting one element tensor as the output, even though I asked for more:
```
Python 3.6.8 |Anaconda, Inc.| (default, Dec 30 2018, 01:22:34) 
[GCC 7.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.FloatTensor(10).cuda().uniform_()
tensor(0.7869, device='cuda:0')
>>> torch.FloatTensor(1000).cuda().uniform_()
tensor(0.7470, device='cuda:0')
>>> torch.FloatTensor(10000).cuda().uniform_()
tensor(0.0427, device='cuda:0')
>>> 
```
Adding out as an input fixed it. @colesbury do you know why?

colesbury replied:
Oh, that sounds like a bug in TensorIterator. The problem is the shape computation, which tries to preserve some legacy resizing behavior and it doesn't work when there's only one Tensor.

The easiest workaround is copy this line to prevent resizing the output. I think that's better than adding the Tensor as an input.

https://github.com/pytorch/pytorch/blob/558c6c4d8ab76f1c8e075a230af584258c9c89e8/aten/src/ATen/native/TensorIterator.cpp#L495

We'll fix the problem in TensorIterator more generally later.

ezyang commented on aten/src/ATen/native/cuda/Distributions.cu:
```diff
@@ -23,17 +25,158 @@
+ * CUDA 10 fixed this problem. However, for backwards compatibility, in the following kernels
+ * we are using curand distributions that utilize curand4 call. curand4 call doesn't have the
+ * register spilling problem.
+ */
```
Nice comments!

ezyang commented on aten/src/THC/generic/THCTensorRandom.cu:
```diff
@@ -191,7 +176,9 @@ void THCTensor_(multinomial)(struct THCState *state,
     THCTensor *sampled = THCTensor_(newWithSize2d)(state, numDist, n_sample);
-    THCTensor_(uniform)(state, sampled, 0.0, 1.0);
+    c10::raw::intrusive_ptr::incref(sampled);
+    auto out = at::Tensor(c10::intrusive_ptr<at::TensorImpl, at::UndefinedTensorImpl>::reclaim(sampled));
```
`THTensor_wrap`

colesbury commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -486,6 +486,13 @@ std::unique_ptr<TensorIterator> TensorIterator::unary_op(Tensor& out, const Tens
 
+std::unique_ptr<TensorIterator> TensorIterator::nullary_op(Tensor& out) {
+  auto builder = TensorIterator::Builder();
+  builder.add_input(out);
```
Remove `builder.add_input(out)`

ezyang commented on aten/src/ATen/test/cuda_distributions_test.cu:
```diff
@@ -0,0 +1,77 @@
+  //   should be at least the number of curand() random numbers used in
+  //   each thread. In this test, we make sure that uniform_ correctly
+  //   increments philox and doesn't reuse randoms from previous calls.
+  //    - We check that by first getting 4 randoms from uniform_.
```
Should we be testing for other numbers of generated randoms, or is that not useful?

syed-ahmed replied:
Yes, that could be helpful. I'll add some more cases.

ezyang commented on aten/src/ATen/native/cuda/Distributions.cu:
```diff
@@ -256,5 +401,48 @@ Tensor& bernoulli_scalar_cuda_(Tensor &self, double p, Generator* gen) {
+      "uniform_ expects to return a [from, to) range, but found from=", from,
+      " > to=", to);
+    AT_CHECK((to - from) <= std::numeric_limits<scalar_t>::max(),
+          "uniform_ expects to-from ≤ std::numeric_limits<double>::max(), but found to=", to,
```
Use of Unicode in source file makes me unhappy

syed-ahmed replied:
Didn't quite get it. So just change the message to `uniform_ expects (to-from) difference to be less than or equal to the maximum finite value representable by the numeric type scalar_t`? :)

syed-ahmed replied:
oh `≤` I see

syed-ahmed replied:
I totally copy-pasted the requirements from here: https://en.cppreference.com/w/cpp/numeric/random/uniform_real_distribution/uniform_real_distribution

colesbury replied:
Also, the error message always says `<double>` but the code may operate on float and half.

ezyang commented on aten/src/ATen/native/cuda/Distributions.cu:
```diff
@@ -256,5 +401,48 @@ Tensor& bernoulli_scalar_cuda_(Tensor &self, double p, Generator* gen) {
+          "uniform_ expects to-from ≤ std::numeric_limits<double>::max(), but found to=", to,
+          " and from=", from, " which result in to-from to exceed the limit");
+
+    using accscalar_t = at::acc_type<scalar_t, true>;
```
ugh, I should have never let us define `acc_type` with a boolean template parameter lol

ezyang commented on aten/src/ATen/native/cuda/Distributions.cu:
```diff
@@ -256,5 +401,48 @@ Tensor& bernoulli_scalar_cuda_(Tensor &self, double p, Generator* gen) {
+    // define lambda to reverse bounds, multiply 'range' and add 'from_'
+    auto uniform_func = [range, from] __device__ (accscalar_t rand) {
+      // reverse the bounds of curand4 from (0, 1] to [0, 1)
+      // Note that this method is from legacy THCTensorRandom and is likely to give
```
Noob question: I thought you were changing the RNG algorithm in this PR, so what's the business with "legacy THCTensorRandom"?

ngimel replied:
Syed is ooto, so filling in for him. It does not matter on what underlying algorithm is used to produce random bits (we are changing it here, from mtgp to philox), you still need a way to transform random bits into a uniformly distributed float. Part of this transformation is done in curand_uniform, that produces (0,1] float/double. In pytorch, however, we need [0,1), and it is slightly tricky to get. Currently (legacy) THCTensorRandom is using a slightly distorting way of doing so (all 1's are turned into 0's), and, in the interest of minimizing changes, we just follow this behavior here.

ezyang commented:
I guess I'll defer to Sam or someone else for review of the TensorIterator-using code, but the rest looks good.

syed-ahmed committed:
Address review changes

syed-ahmed commented:
@ezyang @colesbury I have addressed the review.
- Used THTensor_wrap.
- Fixed typo and error message.
- Added the TensorIterator workaround.
- Added another test case for verifying philox increment.

Also note that this PR: https://github.com/pytorch/pytorch/pull/20298 has to go in first. `__device__` lambdas can't be created in CUDA 8 from `__device__` functions and so the CUDA 8 tests fail for this PR.

ezyang commented:
Is the dependency why Windows is failing?

```
build/aten/src/ATen/native/cuda/Distributions.cu(405): error : On Windows, the enclosing parent function ("uniform_kernel_cuda") for an extended __device__ lambda cannot have internal or no linkage [C:\Jenkins\workspace\caffe2-builds\py2-cuda9.0-cudnn7-windows-build\caffe2\caffe2_gpu.vcxproj]
```

syed-ahmed committed:
[windows] can't have internal linkage for parent functions enclosing __device__ lambdas

syed-ahmed commented:
> Is the dependency why Windows is failing?
> 
> ```
> build/aten/src/ATen/native/cuda/Distributions.cu(405): error : On Windows, the enclosing parent function ("uniform_kernel_cuda") for an extended __device__ lambda cannot have internal or no linkage [C:\Jenkins\workspace\caffe2-builds\py2-cuda9.0-cudnn7-windows-build\caffe2\caffe2_gpu.vcxproj]
> ```

I think this is happening because I made `static void uniform_kernel_cuda(TensorIterator& iter, double from_, double to_, Generator* gen) {` to have internal linkage. Remove the static. Let's see if it builds now.

ezyang commented:
Made it worse lol

syed-ahmed commented:
The floating point exception in the CI is legit. It's happening because of a division by zero coming from `iter.numel()`.
@colesbury Would you happen to know why iter.numel() is zero for
```
import torch
a = torch.rand((0,10), device="cuda", dtype=torch.float32, requires_grad=True)
```
?

soumith commented:
@syed-ahmed I expect because a zero-dimensional Tensor doesn't have any elements

colesbury commented:
@syed-ahmed: I think the bug is here:

https://github.com/pytorch/pytorch/pull/20292/files#diff-95af65e8f86fe4424ff5fd0986ef9c6fR87

syed-ahmed committed:
Fix division by zero bug in distribution_nullary_kernel and make types consistent

syed-ahmed commented:
@colesbury @soumith Thanks! That did it.

houseroad reviewed (commented):
This seems to break the CUDA build:

https://circleci.com/gh/pytorch/pytorch/1674786?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link/console

May 13 17:45:45 /var/lib/jenkins/workspace/aten/src/ATen/native/cuda/Distributions.cu(407): error: An extended __device__ lambda must not be defined in a function that is defined within another function

ngimel commented:
cuda 8 is no longer supported, it was removed in another PR.

colesbury commented:
@houseroad @ezyang That's a CUDA 8 build, but https://github.com/pytorch/pytorch/pull/20298 looks like it removed CUDA 8?

ezyang commented:
It looks like we didn't delete the relevant Caffe2 job, and this is reflected in the PR status. My bad.
